File: AoC2019/JavaScript/day16.js
//puzzle setup
const fs = require('fs');
var input = "";
const phaseMax = 100;

var input = fs.readFileSync('../day16in.txt', 'utf8');
var parsed = input.split('');
//console.log(arr);

//do stuff?
var pattern = [];
var phaseCurr = 1;
var phaseHolder = [];
var temp = 0;
var phaseIdx = 1;
var debugCount = 0;

for (let i = 0; i < 100; i++)
{
    debugCount = i;
    phaser(parsed);
}

function phaser(arr)
{
    arr = [...arr];
    for (let i = 0; i < arr.length; i++)
    {
        makePattern(i+1);
        for (let j = 0; j < arr.length; j++)
        {
            temp += pattern[phaseIdx] * arr[j];
            phaseIdx++;
            if (phaseIdx === pattern.length) phaseIdx = 0;
        }
        if (String(temp).length >= 2) temp = Math.abs(temp) % 10;
        phaseHolder.push(temp);
        temp = 0;
        phaseIdx = 1;
        pattern = [];
        //console.log(phaseHolder);
    }
    parsed = phaseHolder;
    phaseHolder = [];
    /*if (debugCount == 99) */console.log("iteration " + debugCount + ": " + parsed.join(''));
}

function makePattern(times)
{
    for (let i = 0; i < times; i++) pattern.push(0);
    for (let i = 0; i < times; i++) pattern.push(1);
    for (let i = 0; i < times; i++) pattern.push(0);
    for (let i = 0; i < times; i++) pattern.push(-1);
}

File: AoC2016/Python3/day16.py
#puzzle setup
pInput = '01111010110010011'
disk = 272  #input is 35651584 for part 2

#do stuff
checksum = ''

def getChecksum(cs):
    temp = ''
    for i in range(0,len(cs),2):
        if cs[i] == cs[i+1]:
            temp += '1'
        else:
            temp += '0'
    return temp

while len(pInput) < disk:
    pInput = pInput + '0' + pInput[::-1].replace('0','a').replace('1','0').replace('a','1')

checksum = pInput[:disk]

while len(checksum) % 2 == 0:
    checksum = getChecksum(checksum)

print(checksum)